File: 1. 단계별로_풀어보기/30~39/31. 동적 계획법 2/7. 11062.py
import sys

sys.setrecursionlimit(10000)

input = sys.stdin.readline
output = sys.stdout.write

T = int(input())

# 글로벌 변수처럼 쓸 변수
accArr = None
cardArr = None
resultArr = None

def getResult(start, end):
  if resultArr[start][end] > 0:
    return resultArr[start][end]

  if start == end:
    resultArr[start][end] = cardArr[start]
    return cardArr[start]

  rangeSum = accArr[end + 1] - accArr[start]

  leftChoose = rangeSum - getResult(start + 1, end)
  rightChoose = rangeSum - getResult(start, end - 1)

  resultArr[start][end] = max(leftChoose, rightChoose)
  return resultArr[start][end]

for _ in range(T):
  N = int(input())
  cardArr = list(map(int, input().split()))
  
  accArr = list(0 for i in range(N + 1))
  for i in range(1, N + 1):
    accArr[i] = accArr[i - 1] + cardArr[i - 1]
  
  resultArr = list(list(0 for i in range(N + 1)) for j in range(N + 1))

  output(f"{getResult(0, N - 1)}\n")

File: 1. 단계별로_풀어보기/30~39/31. 동적 계획법 2/7. 11062.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
let remain = 0
const T = +inputArr[0]
const queryArr = inputArr
  .slice(1)
  .map((inputRow, rowIdx) => {
    if (remain === 0) {
      const N = +inputArr[rowIdx + 1]
      const cardArr = inputArr[rowIdx + 2].trim().split(' ').map(Number)
      remain = 1
      return [N, cardArr]
    } // ::
    else {
      remain--
      return null
    }
  })
  .filter(block => block !== null)

// 쿼리마다 연산 수행
queryArr.forEach(query => {
  const [N, cardArr] = query

  // 카드의 구간합을 구하기 위한 누적합 배열
  const accArr = Array.from({length: N + 1}).fill(0)
  for (let i = 1; i <= N; i++) {
    accArr[i] = accArr[i - 1] + cardArr[i - 1]
  }

  // 어떤 사람이 특정 구간을 접했을때 얻을 수 있는 최대점수
  // resultArr[start][end] : start 부터 end 까지 카드가 남았을때, 얻을 수 있는 점수의 최대값
  const resultArr = Array.from({length: N}, () => Array.from({length: N}).fill(0))

  // resultArr 의 초기조건을 넣어준다.
  for (let i = 0; i < N; i++) {
    resultArr[i][i] = cardArr[i]
  }

  /**
   * Bottom-UP 연산을 수행한다.
   *
   * 특정 구간의 길이에 대해서, 각 시작점마다 어느 카드를 가지고 오는게 이득인지 연산해나간다.
   */
  for (let len = 2; len <= N; len++) {
    for (let start = 0; start + len - 1 <= N - 1; start++) {
      const end = start + len - 1
      const rangeSum = accArr[end + 1] - accArr[start]

      resultArr[start][end] = Math.max(
        // ::
        rangeSum - resultArr[start][end - 1],
        rangeSum - resultArr[start + 1][end]
      )
    }
  }

  // 출력
  process.stdout.write(`${resultArr[0][N - 1]}\n`)
})
